### ISSUE TITLE
DeclarationReference: Parser does not handle certain inputs correctly

### ISSUE THREAD

**Issue by octogonz** [microsoft/tsdoc#174]
@rbuckton I'm using this issue to track some bugs found while testing the DeclarationReference API.  They are minor enough to be fixed in the same PR.

1. The `@microsoft/rush-stack-compiler-3.5` package name is not handled correctly because of the `.` in the name:
   ```ts
   // (Actual) SyntaxError: SyntaxError: Invalid Component '@microsoft/rush-stack-compiler-3.5'
   // (Expected) @microsoft/rush-stack-compiler-3.5!
   console.log(DeclarationReference.package('@microsoft/rush-stack-compiler-3.5').toString());

   // (Actual) @microsoft/rush-stack-compiler-3.5
   // (Expected) @microsoft/rush-stack-compiler-3.5!
   console.log(DeclarationReference.parse('@microsoft/rush-stack-compiler-3.5!').toString());

   // (Works correctly) @scope/package!
   console.log(DeclarationReference.parse('@scope/package!').toString());
   ```
2. The parser wrongly accepts an NPM package name that is missing the `!`.  The input should really be interpreted as a package, because `@` is not a valid identifier, but packages should require the `!` suffix to avoid confusion:

   ```ts
   // (Actual) @scope/my-package
   // (Expected) SyntaxError
   console.log(DeclarationReference.parse('@scope/my-package').toString());

   // (Works correctly) @scope/my-package!
   console.log(DeclarationReference.parse('@scope/my-package!').toString());
   ```

3. The parser wrongly accepts identifiers containing spaces.
   ```ts
   // (Actual) a.b c.d
   // (Expected) SyntaxError
   console.log(DeclarationReference.parse('a.b c.d').toString());
   ```

**Comment by octogonz** [comment]
@rbuckton Seems like this is not an LR(1) problem.  The current TSDoc parser scans forwards looking for the package delimiter (`#` instead of `!`) like this:

https://github.com/microsoft/tsdoc/blob/79c551169f0fbb3a2017233f4f1361d1e8ccd92b/tsdoc/src/parser/NodeParser.ts#L927-L992

Even if you could somehow represent this as LR(1), seems like the error messages may be better if the algorithm can tell if something looks like an NPM package or not.

**Comment by rbuckton** [comment]
@octogonz `@microsoft/rush-stack-compiler-3.5` may not work, but have you tried `"@microsoft/rush-stack-compiler-3.5"`? Since it includes a `.` it probably needs to be quoted. I can look at improving this when time allows.

**Comment by octogonz** [comment]
I might be able to work on it this weekend. Do you have any objection to simply scanning forwards until we find the `!` delimiter and then parsing the two halves separately? That's how the old parser works. I don't know how to represent it in your grammar, though.

Quoted NPM package names doesn't seem like a good option, since package names and import paths never need escapes in practice. They are much better behaved than TypeScript identifiers (which e.g. may be forced to match a freeform string key in an interface  for a JSON object).

**Comment by rbuckton** [comment]
> Do you have any objection to simply scanning forwards until we find the ! delimiter and then parsing the two halves separately?

No, its probably not a problem. I took a look and should have a PR up shortly.

**Comment by rbuckton** [comment]
@octogonz: #177 contains the related PR.

**Comment by octogonz** [comment]
@rbuckton I validated that issues (1) and (2) are fixed, but issue (3) still repros:
```ts
// (Actual) a.b c.d
// (Expected) SyntaxError
console.log(DeclarationReference.parse('a.b c.d').toString());
```

I'm going to go ahead and merge your PR, though, so we can unblock https://github.com/microsoft/web-build-tools/pull/1406.  I'll open a separate issue for this (relatively minor) edge case.
